core/scale/detail/variant.hpp

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_SCALE_VARIANT_HPP
+#define KAGOME_SCALE_VARIANT_HPP
+
+#include <boost/variant.hpp>
+#include <outcome/outcome.hpp>
+#include "common/visitor.hpp"
+#include "scale/outcome_throw.hpp"
+#include "scale/scale_error.hpp"
+
+namespace kagome::scale::detail {
+  namespace impl {
+    template <uint8_t i, class F, class H, class... T>
+    void for_each_apply_impl(F &f);
+
+    template <class F, class H>
+    void apply_once(F &f, uint8_t i) {
+      f.template apply<H>(i);
+    }
+
+    template <uint8_t i, class F>
+    void for_each_apply_impl(F &f) {
+      // do nothing, end of recursion
+    }
+
+    template <uint8_t i, class F, class H, class... T>
+    void for_each_apply_impl(F &f) {
+      apply_once<F, H>(f, i);
+      for_each_apply_impl<i + 1, F, T...>(f);
+    }
+
+    template <class Stream, class... T>
+    struct VariantEncoder {
+      const boost::variant<T...> &v_;
+      Stream &s_;
```

Comment:
Fields of struct do not require underscore

Reply:
Made them classes.